File: javascript/merge-sort.js
function mergeSort(arr) {
    let arrLen = arr.length;
    if (arrLen < 2) return arr;

    // split array to two part
    let middle = arrLen/2;
    let left = arr.slice(0, middle);
    let right = arr.slice(middle);

    // recursion for each part until array lower than 2
    left = mergeSort(left);
    right = mergeSort(right);

    // magic happen in merge function
    return merge(left, right);
}

function merge(left, right) {
    let leftIdx = 0;
    let rightIdx = 0;
    let leftLen = left.length;
    let rightLen = right.length;
    let result = [];

    // loop for comparasion whenever left and right array index not reach the end
    while(leftIdx < leftLen && rightIdx < rightLen) {
        if (left[leftIdx] <= right[rightIdx]) {
            result.push(left[leftIdx]);
            leftIdx++;
        } else {
            result.push(right[rightIdx]);
            rightIdx++;
        }
    }

    // clean up for left part
    while(leftIdx < leftLen) {
        result.push(left[leftIdx]);
        leftIdx++;
    }

    // clean up for right part. 
    // all array must put in the result.
    while(rightIdx < rightLen) {
        result.push(right[rightIdx]);
        rightIdx++;
    }

    return result;
}

console.log(mergeSort([9,8,7,6,5,1,2,3,4]));

File: golang/merge-sort/merge-sort.go
package main

import "fmt"

func mergeSort(arr []int) []int {
	if len(arr) < 2 {
		return arr
	}

	// split array to two part
	middle := len(arr) / 2
	left := arr[:middle]
	right := arr[middle:]

	// recursion for each part until array lower than 2
	left = mergeSort(left)
	right = mergeSort(right)

	// magic happen in merge function
	return merge(left, right)
}

func merge(left []int, right []int) []int {
	leftIdx, rightIdx := 0, 0
	leftLen, rightLen := len(left), len(right)
	var result []int

	// loop for comparasion whenever left and right array index not reach the end
	for leftIdx < leftLen && rightIdx < rightLen {
		if left[leftIdx] <= right[rightIdx] {
			result = append(result, left[leftIdx])
			leftIdx++
		} else {
			result = append(result, right[rightIdx])
			rightIdx++
		}
	}

	// clean up for left part
	for leftIdx < leftLen {
		result = append(result, left[leftIdx])
		leftIdx++
	}

	// clean up for right part.
	// all array must put in the result.
	for rightIdx < rightLen {
		result = append(result, right[rightIdx])
		rightIdx++
	}

	return result
}

func main() {
	result := mergeSort([]int{5, 4, 3, 2, 1, 6, 7, 8, 9, 10})
	fmt.Println("result", result)
}
